### PR TITLE
Sorted metric tags to avoid duplicate prom data with gRPC requests

### LINKED ISSUE #3968: Duplicate metrics generated for GRPC requests hitting seldon predict_raw
## Describe the bug
Currently trying to host a model on seldon, using python seldon-core inside kubernetes deployment, I started noticing that the RPS recorded is always coming out to be half of the actual value, on further analysis saw that same prometheus metrics with same tags is generated twice but with different value. 
My metric function looks like this
```
return [
            {
                "type": "TIMER",
                "key": "seldon_run_time_milisecs_histogram",
                "value": self.request_duration_ms,
                "tags": {"service": self.model_name, 'route': <REST or GRPC>},
            },
        ]
```
The duplicate metrics for count with two different values
```
seldon_run_time_milisecs_histogram_count{SAME TAGS, route="GRPC", worker_id="86"} 994
seldon_run_time_milisecs_histogram_count{SAME TAGS, route="GRPC", worker_id="86"} 1005

```

## To reproduce
Host a seldon model and send a curl (any http load testing tool) and grpcurl (or ghz) command, grpc metrics will have two different values while rest looks fine. 

## Expected behaviour
Would like to get an aggregate metric
```
seldon_run_time_milisecs_histogram_count{SAME TAGS, route="GRPC", worker_id="86"} 1999

```

### PR DESCRIPTION
**Pull request by harshita-meena**
**What this PR does / why we need it**:
To ensure no duplication of metrics occur in prometheus due to non-unique temp keys dependent on "tags" of custom metrics.

**Which issue(s) this PR fixes**:

Fixes #3968 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by seldondev** [comment]
Hi @harshita-meena. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) or [todo](https://github.com/orgs/todo/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/todo/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **jklaise**
You can assign the PR to them by writing `/assign @jklaise` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by cliveseldon** [comment]
@harshita-meena Can you look at the linting and test errors?

**Comment by seldondev** [comment]
@harshita-meena: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.

<details>

In response to [this](https://github.com/SeldonIO/seldon-core/pull/4006#issuecomment-1081022189):

>/test this


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

File: python/seldon_core/seldon_methods.py
Comment (harshita-meena) [reply python/seldon_core/seldon_methods.py:65]: Sorry for the confusion, I am testing it locally, it is a list. Ill have to iterate on all the elements of the list to sort the tags.

**Comment by seldondev** [comment]
@harshita-meena: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 48e4a11ba6639f3582a0e1ed3cee77782bdd9108 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4006-integration-nh52g) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by axsaucedo** [comment]
Btw just as a heads up @harshita-meena I had to add a dummy commit as it seems an issue in github was preventing the gh action pipelines to run, but that seems to have done the job

**Comment by axsaucedo** [comment]
@harshita-meena it seems there are some failures on the tests, are you able to run these locally? You can do so with `act` command

**Comment by harshita-meena** [comment]
> @harshita-meena it seems there are some failures on the tests, are you able to run these locally? You can do so with `act` command

I was trying to follow the readme [here](https://github.com/SeldonIO/seldon-core/tree/master/testing/scripts) but didn't get to complete the test, ill try your suggestion. thanks!

**Comment by harshita-meena** [comment]
I ran the act command locally, but seems to be unrelated to my change, I ran again from main branch, which seems to be failing as well. Do you think it is something related to how my env is setup ? is there a particular python version i should use
```
| ________________ ERROR collecting adserver/tests/test_server.py ________________
| ImportError while importing test module '/home/hmeena/development/test/seldon-core/components/alibi-detect-server/adserver/tests/test_server.py'.
| Hint: make sure your test modules/packages have valid Python names.
| Traceback:
| /opt/conda/lib/python3.7/importlib/__init__.py:127: in import_module
|     return _bootstrap._gcd_import(name[level:], package, level)
| adserver/tests/test_server.py:1: in <module>
|     from adserver.server import Protocol, CEServer, CEModel
| adserver/server.py:11: in <module>
|     from adserver.base import CEModel, ModelResponse
| adserver/base/__init__.py:2: in <module>
|     from .alibi_model import AlibiDetectModel
| adserver/base/alibi_model.py:4: in <module>
|     from alibi_detect.utils.saving import load_detector, Data
| /opt/conda/lib/python3.7/site-packages/alibi_detect/utils/saving.py:24: in <module>
|     from alibi_detect.od import (IForest, LLR, Mahalanobis, OutlierAE, OutlierAEGMM, OutlierProphet,
| /opt/conda/lib/python3.7/site-packages/alibi_detect/__init__.py:1: in <module>
|     from . import ad, cd, models, od, utils
| /opt/conda/lib/python3.7/site-packages/alibi_detect/od/__init__.py:10: in <module>
|     from .llr import LLR
| /opt/conda/lib/python3.7/site-packages/alibi_detect/od/llr.py:13: in <module>
|     from alibi_detect.utils.perturbation import mutate_categorical
| /opt/conda/lib/python3.7/site-packages/alibi_detect/utils/perturbation.py:5: in <module>
|     import cv2
| /opt/conda/lib/python3.7/site-packages/cv2/__init__.py:8: in <module>
|     from .cv2 import *
| E   ImportError: libGL.so.1: cannot open shared object file: No such file or directory
| =========================== short test summary info ============================
| ERROR adserver/tests/test_ad_model.py
| ERROR adserver/tests/test_cd_model.py
| ERROR adserver/tests/test_cm_model.py
| ERROR adserver/tests/test_od_model.py
| ERROR adserver/tests/test_server.py
| !!!!!!!!!!!!!!!!!!! Interrupted: 5 errors during collection !!!!!!!!!!!!!!!!!!!!
```

**Comment by axsaucedo** [comment]
@harshita-meena it seems the docker image that you are using to run these tests doesn't have the dependencies required for Python CV2 library - you may need to specify explicitly the base image you use to run, alternatively you will have to extend the image you are using to have the cv2 dependencies (ie `RUN apt-get install ffmpeg libsm6 libxext6  -y`).

What is the command that you are using to run these tests? Are you using the `-P` parameter to override the default image?

**Comment by harshita-meena** [comment]
> @harshita-meena it seems the docker image that you are using to run these tests doesn't have the dependencies required for Python CV2 library - you may need to specify explicitly the base image you use to run, alternatively you will have to extend the image you are using to have the cv2 dependencies (ie `RUN apt-get install ffmpeg libsm6 libxext6 -y`).
> 
> What is the command that you are using to run these tests? Are you using the `-P` parameter to override the default image?

I am just using simple `act`, make sense what you said. Ill give it a shot.

**Comment by harshita-meena** [comment]
I am guessing [this](https://github.com/SeldonIO/seldon-core/blob/master/.github/workflows/python_tests.yml#L12) is the image used by default.

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **majolo**
You can assign the PR to them by writing `/assign @majolo` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
@harshita-meena: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.

<details>

In response to [this](https://github.com/SeldonIO/seldon-core/pull/4006#issuecomment-1116959628):

>/retest


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by harshita-meena** [comment]
@axsaucedo How do i trigger the PR tests ? I created the custom image using Dockerfile with content. 
```
from nektos/act-environments-ubuntu:18.04

RUN apt-get update && apt-get install ffmpeg libsm6 libxext6 -y
```
And ran act with ./act -P hmeena-custom-act, it still seems to fail with the same reason locally. 
I tried running PR tests to check any errors that I might be because of my changes.

**Comment by RafalSkolasinski** [comment]
The related issue seems to be closed now https://github.com/SeldonIO/seldon-core/issues/3968 - do we still need it? Order of items in dict is not guaranteed in Python (the fact that Python 3 preserves order of items in `dict` is the effect of current implementation but not guaranteed behavior -> one would have to use `OrderedDict`). 

@harshita-meena if the https://github.com/SeldonIO/seldon-core/issues/3968 is still the problem it should be re-opened -> if that is the case, what would be the minimal example that illustrates the issue?

**Comment by harshita-meena** [comment]
@RafalSkolasinski I figured out why the issue was happening so closed the issue, but the  core issue was not resolved, I can let you know the actual problem, If you have suggestion for another solution, happy to be onboard with it. 

The issue is specific to a scenario where someone is using **grpc** with **predict_raw** and has **multiple tags**, only then does prometheus generate duplicate metrics (which lead to incorrect metric collection). This happens because the function [handle_raw_custom_metrics](https://github.com/SeldonIO/seldon-core/blob/master/python/seldon_core/seldon_methods.py#L49) is called everytime such request is processed where specifically the seldon_message_to_json returns an unordered dictionary [(related issue](https://github.com/protocolbuffers/protobuf/issues/3848)). 

Because of this unordered dictionary, tags are also not ordered (they are part of metrics) and they are used as [keys](https://github.com/SeldonIO/seldon-core/blob/41738bf547eef2a50a34bd24f5feadd957debbea/python/seldon_core/metrics.py#L202) to collect metrics which in turn leads to duplicates if #keys > 1. 

For my usecase I have commented all the tags to avoid running into this bug, but it is not ideal for us.
I can understand the suggestion has been to use predict but for some reason its not possible to use that at the moment.

**Comment by RafalSkolasinski** [comment]
Thanks @harshita-meena, I think I start to understand the issue. In https://github.com/SeldonIO/seldon-core/issues/3968 I see some code samples. Would you be able to provide a minimal complete example (full `Model.py`, used build commands, deployment, example requests) that illustrates the issue fully?

**Comment by harshita-meena** [comment]
@RafalSkolasinski Sure, ill get back to you by tomorrow with all the components you mentioned.

**Comment by harshita-meena** [comment]
@RafalSkolasinski Posted an example use-case with an lgbm model in a [repository](https://github.com/harshita-meena/seldon-lgbm), let me know if you cannot access it. 

I provided a README to host locally, if you run grpcurl multiple times and analyze the metrics endpoint, you will be able to see duplicate metrics (different values for same prom tags ex searching seldon_run_time_histogram_count).

**Comment by RafalSkolasinski** [comment]
Thank you @harshita-meena, as I understand the problem only occurs on GRPC endpoint and not on REST? 
Also, does it appear if you would only have a model with `predict` method without the `predict_raw` one?

**Comment by harshita-meena** [comment]
@RafalSkolasinski Yes the problem is very specific to GRPC with `predict_raw`, it does not happen in REST or in `predict`. I stumbled into it because predict_raw provided many custom changes, and had to look for support of both REST and GRPC where GRPC was showing me very different stats.

**Comment by RafalSkolasinski** [comment]
Thank you @harshita-meena, your example was very helpful. I identified the source of problems to be in this line https://github.com/SeldonIO/seldon-core/blob/83cfb9851bb0e37dbf5b8efc7cbdbcc2b0be87f3/python/seldon_core/metrics.py#L203

As the order of items in `tags` dictionary there is not guaranteed we may have same metric under two different keys indeed!

I believe simple
```
        return "_".join(["-".join(i) for i in sorted(tags.items())])
```
would to the trick.

**Comment by RafalSkolasinski** [comment]
Great! I am pretty sure that knowing the root if issue one could craft an example that also works wrongly for the predict / rest cases

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **rafalskolasinski**
You can assign the PR to them by writing `/assign @rafalskolasinski` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by harshita-meena** [comment]
@RafalSkolasinski Added the changes, is there a way I can trigger PR tests ? I did not quite figure out how to work it locally.

**Comment by RafalSkolasinski** [comment]
Just unit tests should be sufficient here, however, would you be able to add an unit tests for it?

Some model that returns for first request metrics from predict with tags `{"a": "a", "b": "b"}` and for second ``{"b": "b", "a": "a"}`should be fine. Could be a modification to `UserObject` test case here https://github.com/SeldonIO/seldon-core/blob/master/python/tests/test_runtime_metrics_tags.py

**Comment by harshita-meena** [comment]
Yups definitely, will add it, thanks for an example.

**Comment by seldondev** [comment]
@harshita-meena: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | b5b22c5ee74cda19cc4dd62dc90c046d11b0ccf2 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4006-integration-gqxsl) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@harshita-meena: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.

<details>

In response to [this](https://github.com/SeldonIO/seldon-core/pull/4006#issuecomment-1125581599):

>/test integration
>
>


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

**Comment by RafalSkolasinski** [comment]
The python lint job seemed to fail, could you run `fmt` target from `python/Makefile`?

**Comment by harshita-meena** [comment]
> Great! I am pretty sure that now knowing the root of issue one could craft an example that also works wrongly for the predict / rest cases

I have tried these cases but could not recreate it, maybe its still possible. 
For REST the json data always ensured a sorted order, as for predict with gRPC, there was never a scenario where metrics were ingested **from** proto messages (a json "meta" object was converted **to** proto after metrics ingestion before sending gRPC response).  

Either way, this small change should cover all such cases :)

**Comment by harshita-meena** [comment]
The fmt command got stuck so I used black command first as suggested, now ran isort.

**Review by RafalSkolasinski — APPROVED** [review]
Nice, thanks for finding the issue and contributing the fix!

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *[RafalSkolasinski](https://github.com/SeldonIO/seldon-core/pull/4006#pullrequestreview-973007566 "Approved")*

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [RafalSkolasinski]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by harshita-meena** [comment]
@RafalSkolasinski Hi!
Wanted to know when can i expect the new patch version with this change ?

**Comment by RafalSkolasinski** [comment]
Hi, it will come with Seldon Core 1.14 but I don't know the exact release schedule for it. @cliveseldon may know more

**Pull request merged**
